src/main/network/dns.rs

```diff
@@ -30,27 +30,55 @@ struct Record {
 
 #[derive(Debug, PartialEq)]
 pub enum RegistrationError {
-    InvalidAddr,
-    InvalidName,
-    AddrExists,
-    NameExists,
+    BroadcastAddrInvalid,
+    LocalhostAddrInvalid,
+    MulticastAddrInvalid(Ipv4Addr),
+    UnspecifiedAddrInvalid,
+    NameInvalid(String),
+    AddrExists(Ipv4Addr),
+    NameExists(String),
 }
 
 impl Display for RegistrationError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            RegistrationError::InvalidAddr => write!(f, "address is invalid for registration"),
-            RegistrationError::InvalidName => write!(f, "name is invalid for registration"),
-            RegistrationError::NameExists => {
-                write!(f, "a registration record already exists for name")
+            RegistrationError::BroadcastAddrInvalid => write!(
+                f,
+                "broadcast address '{}' is invalid in DNS",
+                Ipv4Addr::BROADCAST
+            ),
+            RegistrationError::LocalhostAddrInvalid => write!(
+                f,
+                "localhost address '{}' is invalid in DNS",
+                Ipv4Addr::LOCALHOST
+            ),
```

Comment:
nit: this may be confusing if the user used 127.0.0.10 (a loopback address which is caught by `is_loopback()` below), but the error message says the address is 127.0.0.1.

Reply:
Fixed (in 69b5a5511c77f689f2f769ffdf65330a518a5bfe)